From e83215d5d22946885fa388d375b12f1b991a43c1 Mon Sep 17 00:00:00 2001
From: Ziyang Huang <hzyitc@outlook.com>
Date: Sun, 8 Sep 2024 16:40:12 +0800
Subject: [PATCH 3/5] remoteproc: qcom_q6v5_mpd: enable clocks

Use managed clock enable so probe failure and remove disable all clocks
before freeing the root remoteproc context.

Signed-off-by: Ziyang Huang <hzyitc@outlook.com>
Signed-off-by: ViKing <viking.yfy@qq.com>
---
--- a/drivers/remoteproc/qcom_q6v5_mpd.c
+++ b/drivers/remoteproc/qcom_q6v5_mpd.c
@@ -78,6 +78,8 @@ struct q6_wcss {
 	phys_addr_t mem_reloc;
 	void *mem_region;
 	size_t mem_size;
+	struct clk_bulk_data *clks;
+	int num_clks;
 	const struct wcss_data *desc;
 	const char **firmware;
 	struct userpd *upd[MAX_UPD];
@@ -757,6 +759,11 @@ static int q6_wcss_probe(struct platform
 	if (ret)
 		goto free_rproc;
 
+	wcss->num_clks = devm_clk_bulk_get_all_enabled(wcss->dev, &wcss->clks);
+	if (wcss->num_clks < 0)
+		return dev_err_probe(wcss->dev, wcss->num_clks,
+				     "failed to acquire and enable clocks\n");
+
 	ret = qcom_q6v5_init(&wcss->q6, pdev, rproc,
 			     WCSS_CRASH_REASON, NULL, NULL);
 	if (ret)
